File: src/main/java/wraith/fabricaeexnihilo/util/Color.java
package wraith.fabricaeexnihilo.util;

import com.google.common.base.Strings;
import com.google.gson.JsonElement;
import com.google.gson.JsonParseException;
import com.google.gson.JsonPrimitive;

public class Color {

    public float r;
    public float g;
    public float b;
    public float a;

    public Color(float r, float g, float b, float a) {
        this.r = r;
        this.g = g;
        this.b = b;
        this.a = a;
    }

    public Color(int color) {
        this(color, true);
    }

    public Color(int color, boolean ignoreAlpha) {
        this(
                (color >> 16 & 255) / 255F,
                (color >> 8 & 255) / 255F,
                (color & 255) / 255F,
                ignoreAlpha ? 1.0f : (color >> 24 & 255) / 255F
        );
    }

    public Color(String hex) {
        this(hexToInt(hex), hex.length() < 8);
    }

    public int toInt() {
        return toIntIgnoreAlpha() | ((int) (a * 255) << 24);
    }

    public int toIntAlphaOne() {
        return toIntIgnoreAlpha() | -16777216;
    }

    public int toIntIgnoreAlpha() {
        return (((int) (r * 255)) << 16) | (((int) (g * 255)) << 8) | ((int) (b * 255));
    }

    public String toHex() {
        return Strings.padStart(Integer.toHexString(toInt()), 8, '0');
    }

    public String toHexNoAlpha() {
        return Strings.padStart(Integer.toHexString(toIntIgnoreAlpha()), 6, '0');
    }

    public int withAlpha() {
        return new Color(r, g, b, a).toInt();
    }

    public int withAlpha(float a) {
        return new Color(r, g, b, a).toInt();
    }

    public int withAlpha(double a) {
        return new Color(r, g, b, (float) a).toInt();
    }

    public static Color fromJson(JsonElement json) {
        if (!(json instanceof JsonPrimitive primitive) || !(primitive.isString() || primitive.isNumber())) {
            throw new JsonParseException("Expected string or number but found " + json + "!");
        }
    
        if (primitive.isString()) {
            return new Color(primitive.getAsString());
        } else {
            return new Color(json.getAsInt());
        }
    }
    
    public static int hexToInt(String hex) {
        var rgb = hex.length() < 8 ? hex : hex.substring(2);
        var a = hex.length() < 8 ? "FF" : hex.substring(0, 2);

        return (Integer.parseInt(a, 16) << 24) | Integer.parseInt(rgb, 16);
    }

    public static Color average(Color left, Color right, float leftWeight) {
        var rightWeight = 1F - leftWeight;
        var r = (float) Math.sqrt(left.r * left.r * leftWeight + right.r * right.r * rightWeight);
        var g = (float) Math.sqrt(left.g * left.g * leftWeight + right.g * right.g * rightWeight);
        var b = (float) Math.sqrt(left.b * left.b * leftWeight + right.b * right.b * rightWeight);
        var a = left.a * leftWeight + right.a * rightWeight;
        return new Color(r, g, b, a);
    }

    public static Color average(Color left, Color right, double leftWeight) {
        return average(left, right, (float)leftWeight);
    }

    public static Color average(Color left, Color right) {
        return average(left, right, 0.5F);
    }

    public static final Color DARK_RED = new Color("AA0000");
    public static final Color RED = new Color("FF5555");
    public static final Color GOLDEN = new Color("FFAA00");
    public static final Color YELLOW = new Color("FFFF55");
    public static final Color DARK_GREEN = new Color("00AA00");
    public static final Color GREEN = new Color("55FF55");
    public static final Color AQUA = new Color("55FFFF");
    public static final Color DARK_AQUA = new Color("00AAAA");
    public static final Color DARK_BLUE = new Color("0000AA");
    public static final Color BLUE = new Color("5555FF");
    public static final Color LIGHT_PURPLE = new Color("FF55FF");
    public static final Color DARK_PURPLE = new Color("AA00AA");
    public static final Color WHITE = new Color("FFFFFF");
    public static final Color GRAY = new Color("AAAAAA");
    public static final Color DARK_GRAY = new Color("555555");
    public static final Color BLACK = new Color("000000");
    public static final Color ORANGE = new Color("FFA500");


    public static final Color IRON = new Color("BF8040");
    public static final Color GOLD = new Color("FFFF00");

    public static final Color ALUMINUM = new Color("BFBFBF");
    public static final Color ARDITE = new Color("FF751A");
    public static final Color BERYLLIUM = new Color("E5ECDD");
    public static final Color BORON = new Color("939393");
    public static final Color COBALT = new Color("3333FF");
    public static final Color COPPER = new Color("FF9933");
    public static final Color LEAD = new Color("330066");
    public static final Color LITHIUM = new Color("EDEDED");
    public static final Color MAGNESIUM = new Color("F8DEF8");
    public static final Color NICKEL = new Color("FFFFCC");
    public static final Color SILVER = new Color("F2F2F2");
    public static final Color TIN = new Color("E6FFF2");
    public static final Color TITANIUM = new Color("BD87CA");
    public static final Color THORIUM = new Color("333333");
    public static final Color TUNGSTEN = new Color("1C1C1C");
    public static final Color URANIUM = new Color("4E5B43");
    public static final Color ZINC = new Color("A59C74");
    public static final Color ZIRCONIUM = new Color("E6E8C7");
}

File: src/main/kotlin/exnihilofabrico/util/Color.kt
package exnihilofabrico.util

import kotlin.math.sqrt

data class Color(val r: Float, val g: Float, val b: Float, val a: Float) {
    constructor(color: Int, ignoreAlpha: Boolean=true): this(
        (color shr 16 and 255).toFloat() / 255f,
        (color shr 8 and 255).toFloat() / 255f,
        (color and 255).toFloat() / 255f,
        if(ignoreAlpha) 1.0f else (color shr 24 and 255) / 255f
    )
    constructor(hex: String): this(hex.toInt(16), hex.length <= 6)

    fun toInt(): Int {
        return toIntIgnoreAlpha() or ((a*255).toInt() shl 24)
    }

    fun toIntAlphaOne(): Int {
        return toIntIgnoreAlpha() or -16777216
    }

    fun toIntIgnoreAlpha(): Int {
        return ((r*255).toInt() shl 16) or ((g*255).toInt() shl 8) or ((b*255).toInt())
    }

    fun toHexNoAlpha(): String {
        return toIntIgnoreAlpha().toString(16)
    }

    companion object {
        fun average(left: Color, right: Color, leftWeight: Float = 0.5f): Color {
            val rightWeight = 1.0f - leftWeight
            val r = sqrt(left.r*left.r*leftWeight + right.r*right.r*rightWeight)
            val g = sqrt(left.g*left.g*leftWeight + right.g*right.g*rightWeight)
            val b = sqrt(left.b*left.b*leftWeight + right.b*right.b*rightWeight)
            val a = left.a*leftWeight + right.a*rightWeight
            return Color(r,g,b,a)
        }
        @JvmStatic
        val DARK_RED = Color("AA0000")
        @JvmStatic
        val RED = Color("FF5555")
        @JvmStatic
        val GOLD = Color("FFAA00")
        @JvmStatic
        val YELLOW = Color("FFFF55")
        @JvmStatic
        val DARK_GREEN = Color("00AA00")
        @JvmStatic
        val GREEN = Color("55FF55")
        @JvmStatic
        val AQUA = Color("55FFFF")
        @JvmStatic
        val DARK_AQUA = Color("00AAAA")
        @JvmStatic
        val DARK_BLUE = Color("0000AA")
        @JvmStatic
        val BLUE = Color("5555FF")
        @JvmStatic
        val LIGHT_PURPLE = Color("FF55FF")
        @JvmStatic
        val DARK_PURPLE = Color("AA00AA")
        @JvmStatic
        val WHITE = Color("FFFFFF")
        @JvmStatic
        val GRAY = Color("AAAAAA")
        @JvmStatic
        val DARK_GRAY = Color("555555")
        @JvmStatic
        val BLACK = Color("000000")
    }
}

object MetalColors {
    val IRON = Color("BF8040")
    val GOLD = Color("FFFF00")

    val ALUMINUM = Color("BFBFBF")
    val ARDITE = Color("FF751A")
    val BERYLLIUM = Color("E5ECDD")
    val BORON = Color("939393")
    val COBALT = Color("3333FF")
    val COPPER = Color("FF9933")
    val LEAD = Color("330066")
    val LITHIUM = Color("EDEDED")
    val MAGNESIUM = Color("F8DEF8")
    val NICKEL = Color("FFFFCC")
    val SILVER = Color("F2F2F2")
    val TIN = Color("E6FFF2")
    val TITANIUM = Color("BD87CA")
    val THORIUM = Color("333333")
    val TUNGSTEN = Color("1C1C1C")
    val URANIUM = Color("4E5B43")
    val ZINC = Color("A59C74")
    val ZIRCONIUM = Color("E6E8C7")
}